controllers/datadogagent/merger/network_policy.go

```diff
@@ -0,0 +1,182 @@
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// NetworkPolicyManager is used to manage network policy resources.
+type NetworkPolicyManager interface {
+	AddKubernetesNetworkPolicy(name, namespace string, podSelector metav1.LabelSelector, policyTypes []netv1.PolicyType, ingress []netv1.NetworkPolicyIngressRule, egress []netv1.NetworkPolicyEgressRule) error
+	BuildKubernetesNetworkPolicy(dda metav1.Object, componentName v2alpha1.ComponentName) error
+}
+
+// NewNetworkPolicyManager returns a new NetworkPolicyManager instance
+func NewNetworkPolicyManager(store dependencies.StoreClient) NetworkPolicyManager {
+	manager := &networkPolicyManagerImpl{
+		store: store,
+	}
+	return manager
+}
+
+// networkPolicyManagerImpl is used to manage network policy resources.
+type networkPolicyManagerImpl struct {
+	store dependencies.StoreClient
+}
+
+// AddKubernetesNetworkPolicy creates or updates a kubernetes network policy
+func (m *networkPolicyManagerImpl) AddKubernetesNetworkPolicy(name, namespace string, podSelector metav1.LabelSelector, policyTypes []netv1.PolicyType, ingress []netv1.NetworkPolicyIngressRule, egress []netv1.NetworkPolicyEgressRule) error {
+	obj, _ := m.store.GetOrCreate(kubernetes.NetworkPoliciesKind, namespace, name)
+	policy, ok := obj.(*netv1.NetworkPolicy)
+	if !ok {
+		return fmt.Errorf("unable to get from the store the NetworkPolicy %s", name)
+	}
+
+	policy.Spec.PodSelector = podSelector
+	policy.Spec.PolicyTypes = append(policy.Spec.PolicyTypes, policyTypes...)
+	policy.Spec.Ingress = append(policy.Spec.Ingress, ingress...)
+	policy.Spec.Egress = append(policy.Spec.Egress, egress...)
+	m.store.AddOrUpdate(kubernetes.NetworkPoliciesKind, policy)
+
+	return nil
+}
+
+// BuildAgentKubernetesNetworkPolicy creates the base node agent kubernetes network policy
+func (m *networkPolicyManagerImpl) BuildKubernetesNetworkPolicy(dda metav1.Object, componentName v2alpha1.ComponentName) error {
+	policyName, podSelector := getPolicyMetadata(dda, componentName)
+	ddaName := dda.GetName()
+	ddaNamespace := dda.GetNamespace()
+
+	policyTypes := []netv1.PolicyType{
+		netv1.PolicyTypeIngress,
+		netv1.PolicyTypeEgress,
+	}
+
+	var egress []netv1.NetworkPolicyEgressRule
+	var ingress []netv1.NetworkPolicyIngressRule
+
+	switch componentName {
+	case v2alpha1.NodeAgentComponentName:
+		// The agents are susceptible to connect to any pod that would
+		// be annotated with auto-discovery annotations.
+		//
+		// When a user wants to add a check on one of its pod, they
+		// need to
+		// * annotate its pod
+		// * add an ingress policy from the agent on its own pod
+		// In order to not ask end-users to inject NetworkPolicy on the
+		// agent in the agent namespace, the agent must be allowed to
+		// probe any pod.
+		egress = []netv1.NetworkPolicyEgressRule{
+			{
+				Ports: append([]netv1.NetworkPolicyPort{}, ddIntakePort()),
+			},
+		}
+		ingress = []netv1.NetworkPolicyIngressRule{}
+	case v2alpha1.ClusterAgentComponentName:
```

Comment:
Can we also make sure cross Cluster Agent communication works, similar to how it was introduced in our helm chart https://github.com/DataDog/helm-charts/pull/641